download/download.go

```diff
@@ -17,23 +26,155 @@ Download: Downloads files from the Sensitive Data Archive (SDA). If a directory
           download the urls_list.txt file, and then download all files in this
           list. If file urls are given, they will be downloaded as-is.
 `
+
+// ArgHelp is the suffix text that will be displayed after the argument list in
+// the module help
 var ArgHelp = `
   [urls]
         all flagless arguments will be used as download urls.`
 
+// Args is a flagset that needs to be exported so that it can be written to the
+// main program help
 var Args = flag.NewFlagSet("download", flag.ExitOnError)
 
-// Main download function
-func Download(args []string) {
+// Gets the file name for a URL, using regex
+func getFileNameFromURL(file string) (fileName string, err error) {
+	// Create the file path according to the way files are stored in S3
+	// The folder structure comes after the UID described in the regex
+	re := regexp.MustCompile(`(?i)[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/(.*)`)
+	match := re.FindStringSubmatch(file)
+	if match == nil || len(match) < 1 {
+		return fileName, fmt.Errorf("failed to parse url for downloading file")
+	}
+	fileName = match[1]
+
+	var filePath string
+	if strings.Contains(fileName, "/") {
```

Comment:
I would probably move this folder creation stuff outside the function, as it's not obvious that this happens from the function name.

Reply:
Since they are needed for the folder creation, I thought it makes sense to take them away from the main function. So the function now is getting the url, exports the file and creates the folder where the file will be stored. But if the consensus is to move it, I am fine with it.